lib/Service/EndpointService.php

```diff
@@ -996,6 +1003,195 @@ private function processMappingRule(Rule $rule, array $data): array
+     *
+     * @throws ContainerExceptionInterface
+     * @throws NotFoundExceptionInterface
+     */
+    private function processAuditTrailRule(Rule $rule, Endpoint $endpoint, array $data, string $objectId): array|Response
+    {
+
+        $pathParameters = $this->getPathParameters(endpointArray: $endpoint->getEndpointArray(), path: $data['path']);
+
+        if(isset($pathParameters['audittrail-id']) === true) {
+            $auditrule = $this->objectService->getOpenRegisters()->getPaginatedAuditTrail($objectId, requestParams: ['uuid' => $pathParameters['audittrail-id']]);
+
+            if(count($auditrule) === 1) {
+                $data['body'] = $auditrule[0];
+                return $data;
+            }
+
+            return new JSONResponse(data: ['error' => 'Not found', 'reason' => 'The resource you are looking for does not exist'], statusCode: HTTP::STATUS_NOT_FOUND);
+
+        }
+        $audittrail = $this->objectService->getOpenRegisters()->getPaginatedAuditTrail($objectId);
+
+        $data['body'] = $audittrail['results'];
+
+
+        return $data;
+    }
+
+    /**
+     * Process a locking rule, either locking or unlocking a resource.
+     *
+     * @param Rule $rule Rule containing configuration for the execution of the rule.
+     * @param array $data The data to update
+     * @param string $objectId The object id of the object to lock or unlock
+     *
+     * @return array The updated data array.
+     *
+     * @throws ContainerExceptionInterface
+     * @throws NotFoundExceptionInterface
+     * @throws \OCP\Files\NotFoundException
+     */
+    private function processLockingRule(Rule $rule, array $data, string $objectId): array
+    {
+        $config = $rule->getConfiguration();
+
+        if($config['locking']['function'] === 'lock') {
+            $process = (Uuid::v4())->jsonSerialize();
+            $object = $this->objectService->getOpenRegisters()->lockObject(identifier: $objectId, process: $process, duration: $config['locking']['duration'] ?? 3600);
+        } else if ($config['locking']['function'] === 'unlock') {
+            $object = $this->objectService->getOpenRegisters()->unlockObject(identifier: $objectId);
+        }
+
+        $data['body'] = $this->objectService->getOpenRegisters()->renderEntity(entity: $object->jsonSerialize());
+
+        return $data;
+    }
+
+    /**
+     * Process a rule to write files.
+     *
+     * @param Rule $rule The rule to process.
+     * @param array $data The data to write.
+     * @param string $objectId The object to write the data to.
+     * @param int $registerId The register the object is in.
+     * @param int $schemaId The schema the object is in.
+     *
+     * @return array
+     * @throws ContainerExceptionInterface
+     * @throws NotFoundExceptionInterface
+     * @throws Exception
+     */
+    private function processWriteFileRule(Rule $rule, array $data, string $objectId): array
```

Comment:
Was there not already a write file rule ?

Reply:
That's a design choice Ruben and I made before, it will end up in one RuleService with the next big refactor

Reply:
CHeck